src/rdpc/analysesProcessor.ts

```diff
@@ -215,6 +310,117 @@ export const getAllMergedDonor = async ({
   return mergedDonors;
 };
 
+export const getAllMergedDonorWithSpecimens = async ({
+  analysesFetcher = fetchAnalysesWithSpecimens,
+  egoJwtManager,
+  studyId,
+  url,
+  config,
+  donorIds,
+}: {
+  studyId: string;
+  url: string;
+  egoJwtManager: EgoJwtManager;
+  donorIds?: string[];
+  config?: {
+    chunkSize?: number;
+    state?: StreamState;
+  };
+  analysesFetcher: typeof fetchAnalysesWithSpecimens;
+}): Promise<SpecimensByDonors> => {
+  let mergedDonors: SpecimensByDonors = {};
+
+  if (donorIds) {
+    for (const donorId of donorIds) {
+      logger.info(`streaming analyses with Specimens for donor ${donorId}`);
+      const stream = analysisStream_withSpeicmens({
+        studyId,
+        rdpcUrl: url,
+        egoJwtManager,
+        config,
+        analysesFetcher,
+        donorId,
+      });
+      for await (const page of stream) {
+        if (page.length === 0) {
+          logger.info(
+            `No sequencing experiment analyses with specimens fetched`
+          );
+        }
+        logger.info(
+          `Streaming ${page.length} of sequencing experiment analyses with specimens...`
+        );
+        const donorPerPage = aggregateSpecimensByDonorId(page);
+        mergeAllPagesSpecimensByDonorId(mergedDonors, donorPerPage);
+      }
+    }
+  } else {
+    const stream = analysisStream_withSpeicmens({
+      studyId,
+      rdpcUrl: url,
+      egoJwtManager,
+      config,
+      analysesFetcher,
+    });
+    for await (const page of stream) {
+      if (page.length === 0) {
+        logger.info(
+          `No sequencing experiment analyses with specimens for streaming`
+        );
+      }
+      logger.info(
+        `Streaming ${page.length} of sequencing experiment analyses with specimens...`
+      );
+      const donorPerPage = aggregateSpecimensByDonorId(page);
+      mergeAllPagesSpecimensByDonorId(mergedDonors, donorPerPage);
+    }
+  }
+  return mergedDonors;
```

Comment:
These methods all use the same pattern of:
1. Get a page of data
2. Merge this data with an accumulator object by removing duplicates
3. Repeat until all data retrieved
4. Return the accumulator

This is resulting in the method `mergeAllPagesSpecimensByDonorId` having side effects, specifically modifying the contents of our accumulator object. Perhaps instead we can simplify this by:
1. Get page of data
2. Add all data from page to our accumulator
3. Repeat until all data retrieved
4. Create a new object from the complete list of data that has filtered duplicates
5. Return the filtered object